### ISSUE TITLE
x86 calling conventions broken for floating point returns

### ISSUE THREAD

**Issue by galenbwill** [Vector35/binaryninja-api#4027]
The x86 lifter currently uses XMM0 as the floating point return register, but the win64 and SYS-V calling conventions specify it should be the full-width ZMM0. [This PR](https://github.com/Vector35/arch-x86/pull/25) addresses this issue, but this breaks some of the unit tests (so it has been un-merged).

However, this may be revealing an issue in the analysis, since the ZMM0 register is apparently being included in the default return (clobber?) register sets for functions.

* The ZMM0 register gets included in the return register set for unknown functions, causing cascading issues:
```
======================================================================
FAIL: test_binary___raw (generated.default.unit.TestBinaryNinja)
----------------------------------------------------------------------
# This is caused by zmm0 being included in the clobber set of the unknown function sub_f00302 being called
<---18 same lines--->
+ Function: f00060 HLIL line: int32_t rax_3
+ Function: f00060 HLIL line: int32_t zmm0_1
+ Function: f00060 HLIL line: rax_3, zmm0_1 = sub_f00302("Hello World!\n")

* This is caused by zmm0 being clobbered by any of the preceding function calls
<---1 same lines--->
- Function: f00060 HLIL line: int32_t var_b8 = 42.4199982f.d
?                                                ^ ---------
+ Function: f00060 HLIL line: int32_t var_b8 = 0x4229ae14
?                                              ++  ^^^^^

* This is a continuation of the zmm0 clobber from above
<---3 same lines--->
- Function: f00060 HLIL line: int32_t var_108 = sub_f00302("Hello World!\n")
?                                               ^^^ --- --------------------
+ Function: f00060 HLIL line: int32_t var_108 = rax_3
?                                               ^^^

* These are the same effects caused by zmm0 being implicitly clobbered by unknown functions as above:
<---33 same lines--->
+ Function: f00060 Instruction: f000ff HLIL->LLIL instruction: None
+ Function: f00060 Instruction: f000ff HLIL->MLIL instruction: None
+ Function: f00060 Instruction: f000ff HLIL->MLILS instruction: []
+ Function: f00060 Instruction: f000ff HLIL->LLIL instruction: None
+ Function: f00060 Instruction: f000ff HLIL->MLIL instruction: None
+ Function: f00060 Instruction: f000ff HLIL->MLILS instruction: []
+ Function: f00060 Instruction: f000ff HLIL->LLIL instruction: rax#6, rcx#6, rdx#4, rsi#5, rdi#8, r8#3, r9#3, r10#3, r11#3, xmm0#3, xmm1#3, xmm2#3, xmm3#3, xmm4#3, xmm5#3, xmm6#3, xmm7#3, xmm8#3, xmm9#3, xmm10#3, xmm11#3, xmm12#3, xmm13#3, xmm14#3, xmm15#3, zmm0#5, mem#12 = call(0xf00302, stack = rsp#2 @ mem#11, params = rdi#7, rsi#4, rdx#3, rcx#5, r8#2, r9#2)
+ Function: f00060 Instruction: f000ff HLIL->MLIL instruction: rax_3#6, zmm0_1#2, mem#3 = 0xf00302("Hello World!\n") @ mem#2
+ Function: f00060 Instruction: f000ff HLIL->MLILS instruction: ['"Hello World!\\n"', '0xf00302', 'rax_3#6, zmm0_1#2, mem#3 = 0xf00302("Hello World!\\n") @ mem#2']
<---5 same lines--->
- Function: f00060 Instruction: f0012d HLIL->LLIL instruction: [rbp#1 - 0xb0 {var_b8}].d = zmm0#3.xmm0.d @ mem#15 -> mem#16
?                                                                                               ^
+ Function: f00060 Instruction: f0012d HLIL->LLIL instruction: [rbp#1 - 0xb0 {var_b8}].d = zmm0#6.xmm0.d @ mem#15 -> mem#16
?                                                                                               ^
- Function: f00060 Instruction: f0012d HLIL->MLIL instruction: var_b8#1 = zmm0#3:0.d
?                                                                               ----
+ Function: f00060 Instruction: f0012d HLIL->MLIL instruction: var_b8#1 = zmm0_1#3
?                                                                             ++
- Function: f00060 Instruction: f0012d HLIL->MLILS instruction: ['42.4199982f', 'var_b8#1 = zmm0#3:0.d', 'zmm0#3:0.d']
?                                                                   ^ -------                     ------------------
+ Function: f00060 Instruction: f0012d HLIL->MLILS instruction: ['0x4229ae14', 'var_b8#1 = zmm0_1#3']
?                                                                 ++  ^^^^^                    ++
<---23 same lines--->
- Function: f00060 Instruction: f00188 HLIL->MLILS instruction: ['"Hello World!\\n"', '0xf00302', 'rax_3#6, mem#3 = 0xf00302("Hello World!\\n") @ mem#2', 'var_108#1 = rax_3#6']
+ Function: f00060 Instruction: f00188 HLIL->MLILS instruction: ['rax_3#6', 'var_108#1 = rax_3#6']
```

* The ZMM0 register appears in extraneous ϕ nodes in MMLIL_SSA:
```
======================================================================
FAIL: test_hlil_arrays (generated.default.unit.TestBinaryNinja)
----------------------------------------------------------------------
# Strangeness here caused apparently by having an extra ϕ node for zmm0, leading to rax escaping from the loop iteration expression, leading to a void function being decompiled as returning rax

<---36 same lines--->
+ Function: 100003730 HLIL line:     rax = a + 1
- Function: 100003730 HLIL line: for (uint64_t a = 0; a u< 3; a = a + 1)
?                                                                  ^^^^
+ Function: 100003730 HLIL line: for (uint64_t a = 0; a u< 3; a = rax)
?                                                                 + ^
+ Function: 100003730 HLIL line: return rax
+ Function: 100003730 HLIL line: uint64_t rax

* Looking at MMLIL_SSA, they diverge here:
  35 @ 100003749  xmm15#1 = ϕ(xmm15#0, xmm15#2)
  36 @ 100003749  zmm0#1 = ϕ(zmm0#0, zmm0#2)
  37 @ 100003749  mem#1 = ϕ(mem#0, mem#2)
  vs
  35 @ 100003749  xmm15#1 = ϕ(xmm15#0, xmm15#2)
  36 @ 100003749  mem#1 = ϕ(mem#0, mem#2)
* For some reason this causes:
   4 @ 100003829  return rax:0
  vs
   4 @ 100003829  return 
* And the outer loop's iter clause:
100003749      uint64_t rax
100003749      for (uint64_t a = 0; a u< 3; a = rax)
10000381b      {
...
100003815          rax = a + 1
100003815      }
100003829      return rax
  vs
100003749      for (uint64_t a = 0; a u< 3; a = a + 1)

* The rest of these are because 100003815 maps to multiple MLILS, whereas it used to map to nothing:
<---11 same lines--->
+ Function: 100003730 Instruction: 100003815 HLIL->LLIL instruction: rax#4 = rax#3 + 1
+ Function: 100003730 Instruction: 100003815 HLIL->MLIL instruction: rax#4 = rax_12#3 + 1
+ Function: 100003730 Instruction: 100003815 HLIL->MLILS instruction: ['1', 'rax#4 = rax_12#3 + 1', 'rax_12#3', 'rax_12#3 + 1']
<---2 same lines--->
- Function: 100003730 Instruction: 10000381b HLIL->MLILS instruction: ['1', 'a#3 = rax#4', 'rax_12#3', 'rax_12#3 + 1']
?                                                                      -----                   --- ^^^^^^^^^^^^^^^^^
+ Function: 100003730 Instruction: 10000381b HLIL->MLILS instruction: ['a#3 = rax#4', 'rax#4']
?                                                                                          ^

* And this is because the return is not void, even though it should be:
<---1 same lines--->
- Function: 100003730 Instruction: 100003829 HLIL->MLIL instruction: return 
+ Function: 100003730 Instruction: 100003829 HLIL->MLIL instruction: return rax#1:0
?                                                                           +++++++
- Function: 100003730 Instruction: 100003829 HLIL->MLILS instruction: ['return ']
+ Function: 100003730 Instruction: 100003829 HLIL->MLILS instruction: ['rax#1:0', 'return rax#1:0']
?                                                                      +++++++++++        +++++++
```

This is available for testing here: https://github.com/Vector35/arch-x86/tree/test_full_width_fp_return_register

**Comment by plafosse** [comment]
I believe this was fixed in 3.6.4705 in conjunction with https://github.com/Vector35/binaryninja-api/issues/3191

**Comment by ConnorBP** [comment]
This issue is affecting me on the return arg. Floating point returns in xmm0 on function pointers with no function body within the same module are impossible to convince that they shouldn't be discarding the xmm0 during the translation from low to medium IL
